x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml

```diff
@@ -313,3 +313,157 @@ setup:
               query: "octopus comforter smells"
 
   - match: { status: 400 }
+
+---
+"Search on a sparse_vector field with dots in the field names":
+
+  - requires:
+      cluster_features: [ "gte_v8.16.0" ]
+      reason: dots in field names allowed starting in in 8.16.0
+
+  - do:
+      indices.create:
+        index: index-with-sparse-vector2
+        body:
+          mappings:
+            properties:
+              ml.tokens:
+                type: sparse_vector
+
+  - match: { acknowledged: true }
+
+  - do:
+      headers:
+        Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
+      index:
+        index: index-with-sparse-vector2
+        id: "has-dots"
+        refresh: true
+        body:
+          ml:
+            tokens:
+              running: 2.4097164
+              good: 2.170997
+              run: 2.052153
+              race: 1.4575411
+              for: 1.1908325
+              5.0k: 2.489943
+
+  - match: { result: "created" }
+
+  - do:
+      headers:
+        Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
+      get:
+        index: index-with-sparse-vector2
+        id: "has-dots"
+
+  - match:
+      _source:
+        ml:
+          tokens:
+            running: 2.4097164
+            good: 2.170997
+            run: 2.052153
+            race: 1.4575411
+            for: 1.1908325
+            5.0k: 2.489943
+
+  - do:
+      search:
+        index: index-with-sparse-vector2
+        body:
+          query:
+            sparse_vector:
+              field: ml.tokens
+              query_vector:
+                5.0k: 2.489943
+
+  - match: { hits.total.value: 1 }
+
+---
+"Search on a nested sparse_vector field with dots in the field names and conflicting child fields":
+
+  - requires:
+      cluster_features: [ "gte_v8.16.0" ]
+      reason: dots in field names allowed starting in in 8.16.0
+
+  - do:
+      indices.create:
+        index: index-with-sparse-vector3
+        body:
+          mappings:
+            properties:
+              parent:
+                type: object
+                subobjects: false
+                properties:
+                  foo:
+                    type: sparse_vector
+                  foo.bar:
+                    type: sparse_vector
+
+  - match: { acknowledged: true }
+
+  - do:
+      headers:
+        Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
+        Content-Type: application/json
+      bulk:
+        index: index-with-sparse-vector3
+        refresh: true
+        body: |
+          {"index": { "_id": "parent-foo" }}
+          {"parent.foo": { "bar.baz": 1.0 }}
+          {"index": { "_id": "parent-foo-bar" }}
+          {"parent.foo.bar": { "baz": 2.0 }}
```

Comment:
what is the behavior if these are provided in the same document?

Reply:
Thanks for calling out this edge case @benwtrent!

Running the following script: 

```
PUT dot-test
{
  "mappings": {
    "properties": {
      "parent": {
        "type": "object",
        "subobjects": false,
        "properties": {
          "foo": {
            "type": "sparse_vector"
          },
          "foo.bar": {
            "type": "sparse_vector"
          }
        }
      }
    }
  }
}

PUT dot-test/_doc/3
{
  "parent.foo": {
    "bar.baz": 1
  },
  "parent.foo.bar": {
    "baz": 2
  }
}

GET dot-test/_search
{
  "query": {
    "sparse_vector": {
      "field": "parent.foo",
      "query_vector": {
        "bar.baz": 1
      }
    }
  }
}

GET dot-test/_search
{
  "query": {
    "sparse_vector": {
      "field": "parent.foo.bar",
      "query_vector": {
        "baz": 2
      }
    }
  }
}
```

The query on `parent.foo` returns results, but the query on `parent.foo.bar` does not. This makes sense due to the way we navigate fields when parsing documents. 

I'm inclined to think there's still value here, despite this edge case. However, I would be amenable to the following: 

1. Adding a property such as `allow_dots` that clearly indicates that we want to support dots in `sparse_vector` field names, defaulting to false. 
2. Documenting this edge case warning people against sub objects with conflicting names.

Reasoning: There's a lot of value to do this, and there are good use cases for wanting to do this (e.g. unit abbreviations). This also feels like an edge case that wouldn't be exercised that often. 

WDYT?

Reply:
I need to think on this more. We need to prevent users from accidentally getting into a situation where their data is just not queryable.

Even if its an edge case, it would be really bad to index a bunch of stuff and just not be able to query it with no error or warning at all.

Reply:
This behavior occurs because we try to [deduplicate](https://github.com/elastic/elasticsearch/blob/2a897d9e607cf4d86ce0fc69101e7e16af61a5aa/server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java#L192) the "fieldName.featureName" pairs to keep the highest score. However, the key we use for deduplication can sometimes clash among features that don't appear in the same field (e.g. [parent.foo].bar.baz and [parent.foo.bar].baz have the same key for dedup).
We use a dedicated Lucene field for each sparse vector field so the clash only appears in our deduplication logic. 
I understand Ben's concern here, as this could be problematic. In such cases, the first "fullPath.feature" wins, and others are discarded, even though they are in different fields.
So instead of adding a new option, we simply need to adapt our deduplication strategy to account for the fact that fullPath+featureName is no longer unique. 
We don't need a new option, just a better [key](https://github.com/elastic/elasticsearch/blob/2a897d9e607cf4d86ce0fc69101e7e16af61a5aa/server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java#L189) ;).